keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,103 @@
+
+It would be easier and expected to allow for changes to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+Relevant Issues:
+
+* https://github.com/kubernetes/kubernetes/issues/71477
+* https://github.com/kubernetes/kubernetes/issues/72198
+
+### Goals
+
+Allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+The apiserver will allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet. Additionally, for each volumeClaimTemplate being expanded, it will be necessary to check if its associated StorageClass has volume expansion enabled. This will be achieved by using an admission controller to check if the associated StorageClasses have `allowVolumeExpansion = true`.
+
+During the StatefulSet update process, the StatefulSet controller will detect an update to a `volumeClaimTemplate` by comparing the updated and current revision of the StatefulSet. This requires the `VolumeClaimTemplates` component of the StatefulSet to be recorded in the StatefulSet's `ControllerRevision` object. 
+
+While updating a pod, the StatefulSet controller will update a referenced persistent volume claim object if its storage request in the associated `volumeClaimTemplate` has been increased.
+
+Not all volumes support online control-plane expansion so this design aims to support both online and offline control-plane volume expansion as follows.
+
+Prerequisite: The StatefulSet controller must be able to determine if a volume supports online expansion.
+
+To accomplish this a new method `RequiresOfflineResize` will be added to the `ExpandableVolumePlugin` interface. Through this it will possible to determine which in-tree volume plugins support online expansion. 
+
+For volumes provisioned through the CSI, an optional boolean field `SupportsOnlineExpansion` will be added into the `CSIDriverSpec` field of the `CSIDriver` object. The `CSIDriver` object stores information about a CSI volume driver deployed on the cluster. If the `CSIDriver` object or the `SupportsOnlineExpansion` field is not registered for a particular volume provisioned through the CSI, the statefulset controller will assume the volume does not support online expansion. Adding this functionality mandates an API change for the mentioned objects.
+
+If a volume is deemed to support online expansion it will be expanded before the pod is terminated for the update. The StatefulSet controller will wait for the file system resize to complete on all such volumes before terminating the associated pod.
```

Comment:
By volume you mean `PVC` right? Can we say PVC will be expanded before terminating the pod? Is there any recovery mechanism we need if a certain replica of stateful set is not running or unreachable?

Reply:
Yes I should be consistent with the terms, using PVC makes more sense here. 

>  Is there any recovery mechanism we need if a certain replica of stateful set is not running or unreachable?

The controller already takes cares of replicas which are not running (i.e recreating them). The expansion functionality will not break that.